pkg/controller/metrics/custom_collectors.go

```diff
@@ -328,18 +334,40 @@ func (cc clusterSyncFailingCollector) Collect(ch chan<- prometheus.Metric) {
 		return
 	}
 
+	clusterDeployments := &hivev1.ClusterDeploymentList{}
+	err = cc.client.List(context.Background(), clusterDeployments)
+	if err != nil {
+		log.WithError(err).Error("error listing cluster deployments")
+		return
+	}
+
 	for _, cs := range clusterSyncList.Items {
 		// Failing cluster syncs
 		cond := controllerutils.FindCondition(cs.Status.Conditions, hiveintv1alpha1.ClusterSyncFailed)
 		if cond != nil && cond.Status == corev1.ConditionTrue {
+			var cdRef hivev1.ClusterDeployment
+			for _, cd := range clusterDeployments.Items {
+				if cd.Namespace == cs.Namespace {
+					cdRef = cd
+					break
+				}
+			}
+			fixedLabels := make(map[string]string, len(cc.dynamicLabels.fixedLabels))
+			fixedLabels["namespaced_name"] = cs.Namespace + "/" + cs.Name
+			if &cdRef != nil {
```

Comment:
- You were correct - I had 2 mistakes here, first was comparing the address instead of the value, and second was to compare the struct directly to nil. After a deep-dive of Laws of Reflection, I believe I have the correct nil/empty comparison now 
- Incorrect, `cdRef` is declared everytime we loop over a cs - aka, it is a temporary variable sparing the length of 1 iteration of clusterSync
- Cool - fixed.
- True, that's why I break

See - usually I would agree, but we're already storing all the clusterdeployments. Do I want to build another map of essentially the same information (I see how it would have less memory because of string and pointers, but it would still have O(n) extra storage), or would I have to loop through them a few times - which is equivalent to the number of failing clustersyncs? I think I prefer the latter.

Reply:
> After a deep-dive of Laws of Reflection, I believe I have the correct nil/empty comparison now

Ack, but see new inline comment.

> * Incorrect, `cdRef` is declared everytime we loop over a cs - aka, it is a temporary variable sparing the length of 1 iteration of clusterSync

Ah, right, my mistake.

> * True, that's why I break
> 
> See - usually I would agree, but we're already storing all the clusterdeployments. Do I want to build another map of essentially the same information (I see how it would have less memory because of string and pointers, but it would still have O(n) extra storage), or would I have to loop through them a few times - which is equivalent to the number of failing clustersyncs? I think I prefer the latter.

As you say, it's not a lot of additional storage if the values are pointers. It's a tradeoff between memory and CPU. Digging in a bit:
In OSD prod, a typical "ns/name" will be in the neighborhood of 64 bytes. A pointer is 8. Let's say for the sake of discussion that the map struct overhead rounds it out to 100. An OSD prod shard has ~600 clusters. So we're talking about building a struct on the order of 60KB via ~600 string copies and pointer assignments.

Now, I did miss that the outer loop is already filtered to clustersyncs that are failing. On p02 right now that's 44/532. Assuming we have to search on average half of the CD list to find a match, that's ~12000 loop iterations with string compares.

So in our little hypothetical, cutting the processing by a factor of ~20 costs us ~60KB. Which is more expensive, the apples or the oranges? :shrug: 

----

But we can actually do better than both:
- Iterate over the ClusterSyncs once, generating a `map[ns/name]{labelValues, seconds}`-ish.
- Iterate over the CDs once, looking for ns/name matches to the above (each of which is O(1) because map) and filling in `labelValues[unreachable]`
- Iterate over the map, pushing the metric to prom (and filling in `unreachable=unspecified` if the CD wasn't found in the previous loop)

This is O(1) -- more specifically {#CSes + #CDs + #failingCSes} -- and in our hypothetical costs something like 4KB of extra memory.